Add Object Size Distributions

noahl opened the pull request:
Add object size distribution support to the api_multistream scenario of
the object_storage_service benchmark. Also introduce a
SizeDistributionIterator class to provide general support for
distributions, and change the object storage upload mechanism from
strings to streams to allow use of APIs that make distribution support
easier.

noahl commented:
@stfeng @yuyantingzero please take a look.

I have verified that the `api_multistream` works on GCP, AWS and Azure (although the AWS and Azure tests were with very, very slightly different versions of the code). In addition, I checked that the `api_data` scenario works on GCP, because that depends on `WriteObjects`, which I changed.

There are also several unit tests for the distribution code itself.

stfeng commented:
Ack. I will take a look at this.

stfeng commented on perfkitbenchmarker/linux_benchmarks/object_storage_service_benchmark.py:
```diff
@@ -339,27 +345,44 @@ def _ProcessMultiStreamResults(raw_result, operation, results, metadata=None):
+      overall_metadata)
+
+  for size in sizes:
+    metadata = metadata.copy()
```
can you make only one copy of the metadata and then overwrite the size inside the for loop?

noahl replied:
I don't think that will work, because each of the Sample object stores a pointer to its metadata dictionary. If I made only one copy, then the metadata for all of the samples would change when I updated inside the loop.

stfeng replied:
SGTM.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -187,32 +187,55 @@ class SizeDistributionIterator(object):
+      self.sizes.append(size)
+      self.cutoffs.append(total_percent)
+
+    # I considered storing probabilities instead of percents, but I'm
```
Consider shortening this comment and remove the usage of "I" from all comments in this file.

noahl replied:
Done. Removed "I" from all comments, and also shortened several of them (including this one).

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -187,32 +187,55 @@ class SizeDistributionIterator(object):
+    # representation and then the total percent is less than 1.0 even
+    # though the percentages add up to 100. I can't prove that this
+    # happens, but storing percentages seems safer.
+    assert total_percent == 100.0
```
if the user pass in a bad distribution, what would the error message look like? Would it be better if we do an if check, and the print out a friendly error message, and then throw exception? (depends on the answer to the first question, if the error message is obviously useful to the user with the current code, then we are fine)

noahl replied:
If the user passes a bad distribution in, it should be caught in the PKB layer by `_DistributionToBackendFormat`, which processes all distributions before sending them to the API test script and has lots of informative error messages. It also checks that the percentages add to 100%, so this assertion should never fail in theory. It's here mostly as a sanity check.

I think it's fine to say that if you want to run the backend worker script yourself, you won't get great error messages (although the message here should still be pretty good). If you think differently, I will add nicer error reporting.

stfeng replied:
SGTM.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -187,32 +187,55 @@ class SizeDistributionIterator(object):
-    raise NotImplementedError('General distributions are not implemented yet.')
-
-  return long(dist.iterkeys().next())
+  return max(dist.iterkeys())
```
Consider saving the max in the initialization function and return the saved value directly from here?

noahl replied:
I'm not sure which initialization function you mean. The API test script initialization? The `SizeDistributionIterator` initialization?

If you mean the `SizeDistributionIterator` initialization, that object itself won't work well because we construct the iterators in the worker threads, but `MaxSizeInDistribution` is called in the parent thread to construct the payload. I think the cleanest way to handle this would be to make an object that represents a distribution. It would have methods to get an iterator and to find the maximum object size in the distribution. I'm just not sure it's worth the extra code it would take to implement that. What would you like me to do?

stfeng replied:
I see, MaxSizeInDistribution is called infrequently, by the parent thread only. (I thought it's called for each and every object, looks like not)

Then this approach of computing max over all keys every time this method is called works just fine.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -354,7 +384,8 @@ def WriteObjects(storage_schema, bucket, object_prefix, count,
 
-  payload_bytes, payload_string = GenerateWritePayload(size)
+  payload = GenerateWritePayload(size)
+  handle = cStringIO.StringIO(payload)
```
have you confirmed whether creation of the handle creates a copy of the payload?

stfeng replied:
Never mind, at this place we are generating the handle only once. should be fine...

noahl replied:
Yes, I confirmed that creating the handle does not copy the payload. I make a 2GB string in a Python process and then made 1000 handles to it. I checked top, and I couldn't see the memory use growing while I was creating the handles.

This handle is only created once, but I do create per-thread handles in `WriteWorker`.

stfeng replied:
SG.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -370,14 +401,14 @@ def WriteObjects(storage_schema, bucket, object_prefix, count,
         latency_results.append(latency)
       if bandwidth_results is not None and latency > 0.0:
         bandwidth_results.append(size / latency)
+      handle.seek(0)
```
wow.... seek is buried here. I'd strongly consider move the seek() to immediately before the place where the handle is going to be read,  that is, inside WriteObjectFromBuffer()

noahl replied:
Done. Moved to `WriteObjectFromBuffer()`.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -387,8 +418,8 @@ def WriteObjectFromBuffer(storage_schema, bucket_name, object_name,
     object_name: Name of the object.
-    payload_bytes: A bytearray to transfer.
-    payload_string: A string version of payload_bytes.
+    stream: A read()-able stream to transfer.
```
Consider adding a condition/constraint here: 

"This function will always seek to the beginning of the stream and write the requested number of bytes from that position to the object storage"

This way, the caller does not have to shoulder the burden of doing seek(0) every time then call this function. (and currently, we don't have a use case to write from a middle of the stream).

noahl replied:
Added.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -405,10 +436,10 @@ def WriteObjectFromBuffer(storage_schema, bucket_name, object_name,
     if host_to_connect is not None:
```
consider adding the seek(0) at here inside this function.

noahl replied:
I added the seek, but moved it to before the timing starts, because I don't think it should be part of the operation latency. Other than that it's exactly like you suggested.

stfeng commented:
I've done my pass, mostly minor comments, thanks for all the unit tests!

Let me know once you are done incorporating my comments, I think we are really close to the finish line here for this one.

meteorfox commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -187,32 +187,53 @@ class SizeDistributionIterator(object):
+    self.cutoffs = []
+
+    total_percent = 0.0
+    for size, percent in dist.iteritems():
```
Order matters here. You should use an `OrderedDict()`.

`dict()` does not guarantee the order of the keys. You could have the case that `next()` returns the incorrect size

Example:
```python
{1024: 10.0, 2048: 45.0, 1048576: 45.0}
```
Then but when iterating over it might give you the following sequence
```python
[(2048, 45.0), (1024, 10.0), (1048576, 45.0)]
```
and now your cutoff list is
```python
[45.0, 55.0, 100.0]
```
Now, say `random.random() * 100.0` returns `9.0`

Then your `next()` function will return `2048` but it should have been `1024`.

stfeng replied:
Such a great catch! Thank you Carlos!!!

noahl replied:
@meteorfox thanks a lot for your review! I really appreciate you taking a look at my code.

I am happy to sort the sizes here. If this ever becomes a general-purpose distribution class, @klausw points out that sorted numbers are a much nicer interface than unsorted. Responding to your comment below, this change will also simplify the unit tests, since we will know exactly which `random()` values correspond to which sizes.

meteorfox replied:
@noahl No problem. Sorting the sizes in ascending order should work as well.

meteorfox commented on script-tests/object_storage_worker_tests.py:
```diff
@@ -15,31 +15,61 @@
+      # We should draw either 1, 10, 1 or 10, 1, 10, but which one
+      # depends on Python's hash function, and I don't want this test
+      # to depend on that.
+      self.assertTrue(values == [1, 10, 1] or values == [10, 1, 10])
```
See comment above about `OrderedDict()`

noahl replied:
I responded above. :)

noahl commented:
@stfeng I pushed an update last night that I believe addresses all of your comments. I also verified that the `api_multistream` and `api_data` scenarios still work on GCP.

@meteorfox, @stfeng I pushed an update just now that responds to Carlos' comment. I verified that `api_multistream` still works on GCP, but I didn't check `api_data` again because the change doesn't affect that code path. Please take a look.

meteorfox commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -187,32 +187,57 @@ class SizeDistributionIterator(object):
+    # possible edge cases where the percentages would add to 100.0 but
+    # their equivalent probabilities would have truncated
+    # floating-point representations and would not add to 1.0.
+    assert total_percent == 100.0
```
Consider adding an error message for when `total_percent != 100.0`, such as users setting percentages that do not add to exactly 100.

noahl replied:
@meteorfox That error should be caught by the PKB driver script, which will print a nice error message if the percentages don't add to 100. This is just a sanity check.

stfeng commented:
LGTM. Please try a run of api-namespace scenario with 100 or 200 iteration before merge. That test case also calls writeobject().

noahl commented:
@stfeng I finished a run of 200 iterations of `api_namespace` overnight. Merging now.
